core/services/remote_access_services.py

```diff
@@ -0,0 +1,128 @@
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module provides interface to communicate with Oppia remotely."""
+
+import hashlib
+import hmac
+import json
+import requests
+
+from core.platform import platform_services
+import vmconf
+
+metadata_services = platform_services.Registry.import_metadata_services()
+
+
+def _get_url():
+    if vmconf.DEV_MODE:
+        return vmconf.DEFAULT_COMMUNICATION_URL
+
+    return vmconf.SERVER_COMMUNICATION_URL
+
+
+def _get_port():
+    if vmconf.DEV_MODE:
+        return vmconf.DEFAULT_COMMUNICATION_PORT
+
+    return vmconf.SERVER_COMMUNICATION_PORT
+
+
+def _get_vm_id():
+    if vmconf.DEV_MODE:
+        return vmconf.DEFAULT_VM_ID
+
+    # Get VMID dynamically from metadata.
+    return metadata_services.get_metadata_param(
+        vmconf.METADATA_VM_ID_PARAM_NAME)
+
+
+def _get_shared_secret():
+    if vmconf.DEV_MODE:
+        return vmconf.DEFAULT_VM_SHARED_SECRET
+
+    # Get shared secret dynamically from metadata.
+    return metadata_services.get_metadata_param(
+        vmconf.METADATA_SHARED_SECRET_PARAM_NAME)
+
+
+def generate_signature(data):
+    """Generates digital signature for given data.
+
+    Args:
+        data: dict. A dictionary data to be transferred over network.`
+    """
+    msg = json.dumps(data)
+    key = _get_shared_secret()
+
+    # Generate signature and return it.
+    return hmac.new(key, msg, digestmod=hashlib.sha256).hexdigest()
+
+
+def fetch_next_job_request():
+    """Returns the next job request to be processed.
+
+    Returns:
+        dict. A dict retrieved remotely from database containing
+        job request data.
+    """
+
+    request_url = "%s:%s/%s" % (
+        _get_url(), _get_port(), vmconf.FETCH_NEXT_JOB_REQUEST_HANDLER)
+
+    payload = {
+        'vm_id': _get_vm_id()
+    }
+    signature = generate_signature(payload)
+    payload['signature'] = signature
+    response = requests.get(request_url, params=payload)
+    return response.json()
+
+
+def store_trained_classifier_model(classifier_data):
+    """Stores the result of processed job request.
+
+    Args:
+        classifier_data: dict. A dictionary containing result of training
+            of classifier.
+
+    Returns:
+        response: response object containing the server's response.
+
+    Raises:
+        Exception: classifier_data is not of dict type.
+        Exception: classifier_data does not contain 'job_request_id' key.
+        Exception: classifier_data does not contain 'training_result' key.
+    """
+
+    # Make sure that classifier_data is in proper foramt.
```

Comment:
perhaps wrap this up in a re-usable validate() method? This validate method would then call the algorithm specific validate() function to validate classifier_data dict,